windows/jenkins_slave.ps1

```diff
@@ -32,10 +32,14 @@ function Check-Call {
     }
 }
 Set-ExecutionPolicy Bypass -Scope Process -Force
-$AutoConnectScript = 'C:\slave-autoconnect.py'
+$AutoConnectScript = 'C:\run-auto-connect.bat'
 $progressPreference = 'silentlyContinue'
-Invoke-WebRequest -Uri https://windows-post-install.s3-us-west-2.amazonaws.com/slave-autoconnect.py -OutFile $AutoConnectScript
+Invoke-WebRequest -Uri https://windows-post-install.s3-us-west-2.amazonaws.com/slave-autoconnect.py -OutFile C:\slave-autoconnect.py
+Invoke-WebRequest -Uri https://windows-post-install.s3-us-west-2.amazonaws.com/run-auto-connect.bat -OutFile $AutoConnectScript
 $trigger = New-ScheduledTaskTrigger -AtStartup -RandomDelay 00:00:30
 $action = New-ScheduledTaskAction -Execute $AutoConnectScript
 $principal = New-ScheduledTaskPrincipal -UserID "NT AUTHORITY\SYSTEM" -LogonType ServiceAccount -RunLevel Highest
 Register-ScheduledTask "JenkinsAutoConnect" -Description "Connect to Jenkins at startup" -Action $action -Trigger $trigger -Principal $principal
+
+# trigger userdata processing on next boot so jenkins slave settings get written
+Check-Call { C:\ProgramData\Amazon\EC2-Windows\Launch\Scripts\InitializeInstance.ps1 –Schedule }
```

Comment:
I don't understand this part.

Reply:
I actually changed the way this works to get it from ec2 tags than from files, but I hadn't push those changes to dev.

Reply:
Could you explain this part?

Reply:
Sure. The userdata is run only on the first instance boot (unless you specify a persist flag in userdata, which we aren't doing.) This command triggers userdata to be executed again on next boot. Since the next boot will be initiated from the Lambda autoscaler and it writes jenkins slave config creation into the userdata, we need to make sure it is triggered again.

Reply:
But aren't the previous lines installing a service that starts on boot? (AtStartup)?

Reply:
Yes, the previous lines are installing the Jenkins runner scripts. The last line schedules handling userdata on next boot (which writes the config files for the jenkins runner.)

Reply:
I see what you are saying, but this is weird, is it only for Windows instances? For Linux I think userdata is executed always when you launch an instance....   am I missing something?